File: DFS/단지번호붙이기.cpp
#include<cstdio>
#include<algorithm>
using namespace std;

int arr[27][27] = { 0, };//맨 밑에 25번째 행 때문에 26번째 행에 0을 넣어줘야함 
int DFS[27][27] = { 0, };
int house[626];
int N;
int temp;

int flood(int i, int j)
{

	if (arr[i][j] == 0 || DFS[i][j] == 1)
		return 0;

	else {

		DFS[i][j] = 1;
		temp++;

		flood(i - 1, j);//위
		flood(i + 1, j);//아래
		flood(i, j - 1);//왼쪽
		flood(i, j + 1);//오른쪽	
	}

	return temp;
}


void solve(int N)
{
	int cnt = 0;

	for (int i = 1; i <= N; i++)
	{
		for (int j = 1; j <= N; j++)
		{
			if (arr[i][j] == 1 && DFS[i][j] == 0)
			{
				temp = 0;
				house[cnt++] = flood(i, j);	//집 수 반환
			}
		}
	}
	printf("%d\n", cnt);//총 단지수

	sort(house, house + cnt);//단지 정렬
	for (int i = 0; i < cnt; i++)
		printf("%d\n", house[i]);
}

int main()
{
	scanf("%d", &N);

	for (int i = 1; i <= N; i++)
		for (int j = 1; j <= N; j++)
			scanf("%1d", &arr[i][j]);

	solve(N);
	return 0;
}

File: DFS/단지번호붙이기.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.Scanner;

public class Danji {

	private static int N;
	private static int [][] arr;
	private static boolean [][] visited;
	private static int danzi;
	private static int cnt;
	private static ArrayList<Integer> list = new ArrayList<Integer>();
	private static String str;
	private static Iterator<Integer> itr;

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		N = sc.nextInt();
		arr = new int[N][N];
		visited = new boolean[N][N];

		for(int i=0;i<N;i++){
			str =sc.next();
			for(int j=0;j<N;j++){
				arr[i][j]=str.charAt(j)-48;
			}
		}
		for(int i=0;i<N;i++){
			for(int j=0;j<N;j++){
				if(arr[i][j]==1 && visited[i][j]==false){
					cnt=0;
					danzi++;
					flood(i,j);
					list.add(cnt);
				}
			}
		}
		Collections.sort(list);
		System.out.println(danzi);

		 itr = list.iterator();
		while(itr.hasNext()){
			System.out.println(itr.next());
		}
	}
	
	public static void flood(int i, int j) {
		if(arr[i][j]==1 && visited[i][j]==false){
			cnt++;
			visited[i][j]=true;
			if(i!=0){
				flood(i-1,j);
			}
			if(j!=0){
				flood(i,j-1);
			}
			if(i!=N-1){
				flood(i+1,j);
			}
			if(j!=N-1){
				flood(i,j+1);
			}
		}
	}
}
